hydra-node/src/Hydra/Chain/Direct/Handlers.hs

```diff
@@ -230,21 +231,24 @@ chainSyncHandler tracer callback getTimeHandle ctx =
                       }
                 }
 
--- | Hardcoded grace time for close transaction to be valid.
--- TODO: make it a node configuration parameter
-closeGraceTime :: SlotNo
-closeGraceTime = 100
-
-fromPostChainTx ::
+prepareTxToPost ::
   (MonadSTM m, MonadThrow (STM m)) =>
   TimeHandle ->
   TinyWallet m ->
   ChainContext ->
   ChainStateType Tx ->
   PostChainTx Tx ->
   STM m Tx
-fromPostChainTx timeHandle wallet ctx cst@ChainStateAt{chainState} tx = do
-  pointInTime <- throwLeft currentPointInTime
+prepareTxToPost timeHandle wallet ctx cst@ChainStateAt{chainState} tx = do
+  pointInTime@(currentSlot, currentTime) <- throwLeft currentPointInTime
+  let (UnsafeContestationPeriod cpNatural) = contestationPeriod ctx
+  -- calculate tx upper bound by using contestation period and current time. See ADR21 for context
+  txEndSlot <- throwLeft $ slotFromUTCTime $ addUTCTime (fromIntegral cpNatural) currentTime
+  -- closeGraceTime determins the tx validity.
+  -- Get the difference between tx upper bound and current slot to determine
+  -- for how many slots this tx is valid.
+  -- NB: Used in 'adjustPointInTime' to calculate tx upper bound.
+  let closeGraceTime = txEndSlot - currentSlot
```

Comment:
Also: We would not need to do `adjustPointInTime` because we would already have the upper bound slot+time here: `addUTCTime (fromIntegral cpNatural) currentTime` is the `UTCTime` and `txEndSlot` is the slot.